File: code/ch15/15.2.2.jump-game.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        n = len(nums)
        dp = [0] * n
        dp[0] = nums[0]
        for i in range(1, n):
            if dp[i - 1] < i:
                dp[i] = dp[i - 1]
            else:
                dp[i] = max(dp[i - 1], i + nums[i])
        return dp[n - 1] >= n - 1

File: code/ch15/15.2.2.jump-game.java
class Solution {
    public boolean canJump(int[] nums) {
        int n = nums.length;
        int[] dp = new int[n];
        dp[0] = nums[0];
        for (int i = 1; i < n; i++) {
            if (dp[i - 1] < i) {
                dp[i] = dp[i - 1];
            } else {
                dp[i] = Math.max(dp[i - 1], i + nums[i]);
            }
        }
        return dp[n - 1] >= n - 1;
    }
}